File: Reference/VB.NET/DynamicGeometry/Interaction/Actions/AbstractAction.vb
Imports DynamicGeometry.Actions

Namespace Actions

	Friend MustInherit Class AbstractAction

		Implements IAction

		Protected MustOverride Sub ExecuteCore()
		Protected MustOverride Sub UnExecuteCore()

		'====================================================================

		Public Sub Execute() Implements IAction.Execute
			If Not CanExecute() Then Return
			ExecuteCore()
			ExecuteCount += 1
		End Sub

		Public Sub UnExecute() Implements IAction.UnExecute
			If Not CanUnExecute() Then Return
			UnExecuteCore()
			ExecuteCount -= 1
		End Sub

		Private mExecuteCount As Integer = 0
		Public Property ExecuteCount() As Integer Implements IAction.ExecuteCount
			Get
				Return mExecuteCount
			End Get
			Set(ByVal Value As Integer)
				mExecuteCount = Value
			End Set
		End Property

		Public Overridable Function CanExecute() As Boolean Implements IAction.CanExecute
			Return ExecuteCount <= 0
		End Function

		Public Overridable Function CanUnExecute() As Boolean Implements IAction.CanUnExecute
			Return Not CanExecute()
		End Function

	End Class
End Namespace

File: Reference/C#/DynamicGeometry/Actions/AbstractAction.cs
namespace GuiLabs.Utils.Actions
{
	public abstract class AbstractAction : IAction
	{
		private int mExecuteCount;
		protected int ExecuteCount
		{
			get
			{
				return mExecuteCount;
			}
			set
			{
				mExecuteCount = value;
			}
		}
		
		public virtual void Execute()
		{
			if (!CanExecute())
			{
				return;
			}
			ExecuteCore();
			ExecuteCount++;
		}

		protected abstract void ExecuteCore();

		public virtual void UnExecute()
		{
			if (!CanUnExecute())
			{
				return;
			}
			UnExecuteCore();
			ExecuteCount--;
		}

		protected abstract void UnExecuteCore();

		public virtual bool CanExecute()
		{
			return ExecuteCount == 0;
		}

		public virtual bool CanUnExecute()
		{
			return !CanExecute();
		}

		/// <summary>
		/// If the last action can be joined with the FollowingAction,
		/// the following action isn't added to the Undo stack,
		/// but rather mixed together with the current one.
		/// </summary>
		/// <param name="FollowingAction"></param>
		/// <returns>true if the FollowingAction can be merged with the
		/// last action in the Undo stack</returns>
		public virtual bool TryToMerge(IAction FollowingAction)
		{
			return false;
		}

		private bool mAllowToMergeWithPrevious = true;
		public bool AllowToMergeWithPrevious
		{
			get
			{
				return mAllowToMergeWithPrevious;
			}
			set
			{
				mAllowToMergeWithPrevious = value;
			}
		}
	}
}
